Update EIP-8037: improve EIP-7702 authorization gas accounting in EIP-8037 (#11715)
ethereum/EIPs: EIPS/eip-8037.md

##### Gas accounting for [EIP-7702](./eip-7702.md) authorizations

> Intrinsic gas charges the worst-case state-gas cost (`(STATE_BYTES_PER_NEW_ACCOUNT + STATE_BYTES_PER_AUTH_BASE) × CPSB`) for each authorization. Because both an authority's account leaf and its 23-byte delegation indicator are written at most once per transaction, even when the same authority appears in multiple authorizations. Adjustments are applied per authorization as the list is processed.

Reviewer: The second phrase reads a bit strange to me. There isa causal clause (i.e., because...), without a main clause (i.e. what is happening because both an authority's account leaf and its 23-byte delegation indicator are written at most once per transaction?)

> Processing maintains a per-tx set `billed` of authorities for which an auth-base creation has already been charged in this transaction. For each **valid** authorization processed in list order, two views of the authority's state are referenced:

Reviewer: ```suggestion
Processing maintains a per-tx set of `billed` authorities for which an auth-base creation has already been charged in this transaction. For each **valid** authorization processed in list order, two views of the authority's state are referenced:
```

- **committed state**: the authority's state at the start of the transaction.

> - **in-flight state**: the authority's state after applying all prior authorizations in this transaction.

Reviewer: Do we assume that the in-flight state is initialized as the committed state?

Reviewer: Also, what is this state exactly? We already have the concept of billed vs non-billed. I think we should clarify this

> **1. Account-creation refill.** If the authority's account leaf already exists in the in-flight state (non-zero nonce, non-zero balance, or non-empty code), refill `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` to `state_gas_reservoir` and refund `ACCOUNT_WRITE` to the refund counter.

Reviewer: I think we should add a phrase before this numbered list. Something like:

```suggestion
Authorization processing follows 3 gas accounting rules:

**1. Account-creation refill.** If the authority's account leaf already exists in the in-flight state (non-zero nonce, non-zero balance, or non-empty code), refill `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` to `state_gas_reservoir` and refund `ACCOUNT_WRITE` to the refund counter.
```

**2. Auth-base refill.** Refill `STATE_BYTES_PER_AUTH_BASE × CPSB` to `state_gas_reservoir` if **any** of:

- the authority was already delegated in the **committed** state
- the authority is in `billed` (a prior authorization in this transaction has already been charged the auth-base creation for this authority)
- `authorization.address` is `ZERO` (the authorization is clearing the delegation; no delegation bytes are written).

> Otherwise, add the authority to `billed`; this authorization is the one charged for the per-tx auth-base creation.

Reviewer: ```suggestion
**3. First charge.** If the authorization does not meet the conditions of rule 1 and 2, add the authority to `billed`; this authorization is the one charged for the per-tx auth-base creation. This charge was already included in the transaction's intrinsic cost.
```

> **3. Cancel a prior auth-base creation.** If `authorization.address` is `ZERO` **and** the authority is in `billed`, additionally refill `STATE_BYTES_PER_AUTH_BASE × CPSB` to `state_gas_reservoir` and remove the authority from `billed`. The chain of authorizations for this authority now writes zero net delegation bytes, so the earlier creation charge is no longer justified. This covers patterns such as `0->a->0`, where the creation by an earlier authorization is undone by a later clearing one in the same list.

Reviewer: ```suggestion
**4. Cancel a prior auth-base creation.** If `authorization.address` is `ZERO` **and** the authority is in `billed`, additionally refill `STATE_BYTES_PER_AUTH_BASE × CPSB` to `state_gas_reservoir` and remove the authority from `billed`. The chain of authorizations for this authority now writes zero net delegation bytes, so the earlier creation charge is no longer justified. This covers patterns such as `0->a->0`, where the creation by an earlier authorization is undone by a later clearing one in the same list.
```

##### Gas accounting for [EIP-7702](./eip-7702.md) authorizations

> Intrinsic gas charges the worst-case state-gas cost (`(STATE_BYTES_PER_NEW_ACCOUNT + STATE_BYTES_PER_AUTH_BASE) × CPSB`) for each authorization. However, an authority's account leaf and its 23 byte delegation indicator can only be written once per transaction, even if the same authority appears in multiple authorizations. Therefore, per authorization adjustments are applied while processing the authorization list.

Reviewer 2: ```suggestion
Intrinsic gas charges the worst-case gas cost (`(STATE_BYTES_PER_NEW_ACCOUNT + STATE_BYTES_PER_AUTH_BASE) × CPSB`) for each authorization. However, an authority's account leaf and its 23 byte delegation indicator can only be written once per transaction, even if the same authority appears in multiple authorizations. Therefore, per authorization adjustments are applied while processing the authorization list.
```

it's not just the worst-case state but `ACCOUNT_WRITE` is also given back

    - **current state**: the account state while processing authorizations, updated after each authorization.

Authorization processing follows 5 gas accounting rules:

**1. Invalid authorizations.** Invalid authorizations are skipped without per-auth processing. Their entire intrinsic state-gas portion, `(STATE_BYTES_PER_NEW_ACCOUNT + STATE_BYTES_PER_AUTH_BASE) × CPSB`, is refilled to `state_gas_reservoir` and refund `ACCOUNT_WRITE` to the refund counter.

**2. Account-creation refill.** If the authority's account leaf already exists in the **current state** (non-zero nonce, non-zero balance, or non-empty code), refill `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` to `state_gas_reservoir` and refund `ACCOUNT_WRITE` to the refund counter.

**3. Auth-base refill.** Refill `STATE_BYTES_PER_AUTH_BASE × CPSB` to `state_gas_reservoir` if **any** of:

- the authority was already delegated in the **pre-transaction state** state
- the authority is in `billed` (a prior authorization in this transaction has already been charged the auth-base creation for this authority)
- `authorization.address` is `ZERO` (the authorization is clearing the delegation)

> **4. First charge.** If the authorization does not meet the conditions of rule 1, 2 and 3, add the authority to `billed`; this authorization is the one charged for the per-tx auth-base creation. This charge was already included in the transaction's intrinsic cost.

Reviewer 2: ```suggestion
**4. First charge.** If the authorization does not meet the conditions of rule 1 and 3, add the authority to `billed`; this authorization is the one charged for the per-tx auth-base creation. This charge was already included in the transaction's intrinsic cost.
```

Depending on rule 2 to be on `billed` breaks the steps below, I don't think it should depend on it. Case for a normal EOA (no code deployed) at the start authorization processing:
- rule 1 doesn't trigger - it's a valid delegation;
- rule 2 triggers, account already exists - STATE_BYTES_PER_NEW_ACCOUNT + ACCOUNT_WRITE refunded;
- rule 3 no trigger - this is a legit delegation as the auth bytes are being set for first time so you are billed for STATE_BYTES_PER_AUTH_BASE;
- rule 4 no trigger - since it was triggered for 2;

Then the next time you loop, in rule 3 if there's a duplicate authorization, you won't see it in `billed` whereas you should. Then you will charge it again.

Author: Good catch!

> **5. Cancel a prior auth-base creation.** If `authorization.address` is `ZERO` **and** the authority is in `billed`, additionally refill `STATE_BYTES_PER_AUTH_BASE × CPSB` to `state_gas_reservoir` and remove the authority from `billed`. The chain of authorizations for this authority now writes zero net delegation bytes, so the earlier creation charge is no longer justified. This covers patterns such as `0->a->0`, where the creation by an earlier authorization is undone by a later clearing one in the same list.

Reviewer: Building on the rule‑4 issue already raised above — rather than fixing how `billed` gets populated, I think the whole set can be dropped. `billed` is reconstructing something that's already in current state: once an earlier authorization delegates `A`, its `codeHash` is non-empty, so "was this authority delegated earlier in the tx?" is just a `codeHash` check. The only thing current state can't distinguish is *pre-tx* vs *in-tx* delegation, which matters solely for the cancel case and is a single read of the pre-tx snapshot.

That collapses the 5 rules to 3. It's equivalent to the intended behavior — the rule‑3 condition `pre_delegated OR billed OR ZERO` is exactly `cur_delegated OR pre_delegated OR ZERO` once `billed = cur_delegated AND NOT pre_delegated` — but reading `codeHash` directly can't have the rule‑4 failure mode, and it fixes `0→a→0` for funded EOAs (which rule 5 misses today because those authorities never enter `billed`).

Concrete suggestion (also drops the `billed` set, fixes the "state state" typo, and tidies rule 1's grammar):

```suggestion
The adjustments enforce a single invariant: the `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` account-leaf portion is charged at most once per authority and only when the account did not exist before the transaction, and the `STATE_BYTES_PER_AUTH_BASE × CPSB` delegation-indicator portion is charged at most once per authority and only when the authority ends the transaction delegated having started it undelegated.

Processing reads the account state at two points:

- **pre-transaction state**: the account state before transaction execution.
- **current state**: the account state while processing authorizations, updated after each authorization. Because an in-transaction delegation leaves a delegation indicator in the authority's `code` field, the current state already records whether an earlier authorization in this transaction delegated the same authority.

For each authorization, let `cur_delegated` be `true` when the authority has a non-empty delegation indicator in the **current state** (`codeHash != emptyHash`), and `pre_delegated` be `true` when it had one in the **pre-transaction state**. Authorization processing follows 3 gas accounting rules:

**1. Invalid authorizations.** Invalid authorizations are skipped without per-auth processing. Their entire intrinsic state-gas portion, `(STATE_BYTES_PER_NEW_ACCOUNT + STATE_BYTES_PER_AUTH_BASE) × CPSB`, is refilled to `state_gas_reservoir` and `ACCOUNT_WRITE` is refunded to the refund counter.

**2. Account-leaf refill.** If the authority's account leaf already exists in the **current state** (non-zero nonce, non-zero balance, or non-empty code), refill `STATE_BYTES_PER_NEW_ACCOUNT × CPSB` to `state_gas_reservoir` and refund `ACCOUNT_WRITE` to the refund counter.

**3. Delegation-indicator refill.**

- If `authorization.address` is not `ZERO` (setting a delegation), refill `STATE_BYTES_PER_AUTH_BASE × CPSB` to `state_gas_reservoir` when `cur_delegated` or `pre_delegated` is `true`. In that case the 23-byte indicator slot is already occupied, so this authorization overwrites it without writing new bytes. Otherwise this authorization performs the net-new creation and its intrinsic charge stands.
- If `authorization.address` is `ZERO` (clearing the delegation), refill `STATE_BYTES_PER_AUTH_BASE × CPSB` to `state_gas_reservoir`, since the clear writes no indicator. Additionally, when `cur_delegated` is `true` and `pre_delegated` is `false`, refill another `STATE_BYTES_PER_AUTH_BASE × CPSB`: the delegation being cleared was created by an earlier authorization in this same transaction, so the chain writes zero net delegation bytes and the earlier creation charge is no longer justified. This covers patterns such as `0->a->0`, where the creation by an earlier authorization is undone by a later clearing one in the same list.
```
